spec/controllers/api/v1/courses_controller_spec.rb

```diff
@@ -0,0 +1,135 @@
+
+    def format_timestamp(timestamp)
+      timestamp.strftime(timestamp_format)
+    end
+
+    # Default sensible params used by tests.
+    let(:params) do
+      {
+        changed_since: format_timestamp(changed_since)
+      }
+    end
+
+    before do
+      allow(controller).to receive(:authenticate)
+
+      # A bit of boilerplate setup since this isn't an integration or
+      # functional test. There may be a more standard way of doing this
+      # which would be better.
+      assigns(:_params).merge! params
+
+      # Stubbing out the data that Rails' url_for relies on is too tricky,
+      # so just stub the whole method out.
+      allow(controller).to receive(:url_for) do |options = {}|
+        'http://test.local/api/v1/course?' + options[:params].to_query
+      end
+      controller.response = response
+    end
+
+    context 'with two courses changed at different times' do
+      let(:old_course)  { create(:course, updated_at: 5.minute.ago.utc) }
+      let(:last_course) { create(:course, updated_at: 1.minute.ago.utc) }
+
+      # We need to define the before block after any let! statements since they
+      # are run in order of definition: we need to call the controller action
+      # after any let! fixtures are created.
+      before do
+        old_course
+        last_course
+
+        controller.index
+      end
+
+      context 'using a changed_since before any courses have changed' do
+        # Gets placed into params.
+        let(:changed_since) { 10.minutes.ago.utc }
+
+        describe 'returned courses in JSON' do
+          subject { response.body }
+
+          it {
+            should have_courses(old_course, last_course)
+          }
+        end
+
+        describe 'generated next link' do
+          subject do
+            # Parse out the query params for testing.
+            Rack::Utils.parse_query(URI(response.headers['Link']).query)
+          end
+
+          its(%w[per_page]) { should eq '100' }
+          its(%w[changed_since]) do
+            should eq format_timestamp(last_course.updated_at)
+          end
+        end
+      end
+
+      context 'using a changed_since after any courses have changed' do
+        describe 'generated next link' do
+          subject do
+            # Parse out the query params for testing.
+            Rack::Utils.parse_query(URI(response.headers['Link']).query)
+          end
+
+          let(:changed_since) { Time.now.utc }
+
+          its(%w[per_page]) { should eq '100' }
+          its(%w[changed_since]) { should eq params[:changed_since] }
+        end
+      end
+    end
+
+    context 'with no courses at all' do
+      # No courses, so this could be anything really.
+      let(:changed_since) { 10.minutes.ago.utc }
+
+      before do
+        controller.index
+      end
+
+      describe 'returned courses in JSON' do
+        subject { response.body }
+
+        # TODO: This isn't very clear, it should be "should_not have_courses"
```

Comment:
?? is this an issue with the matcher?

Reply:
I think there should be some way to get the matcher to behave better in this instance, just haven't looked yet.

Reply:
suggested improvement: https://github.com/DFE-Digital/manage-courses-backend/pull/154/commits/bbbf446a8bd99c36e462044442061a49d3c11918
thoughts @misaka @benilovj  ?